subsys/bluetooth/host/adv.c

```diff
@@ -27,6 +27,112 @@ enum adv_name_type {
 	ADV_NAME_TYPE_SD,
 };
 
+struct bt_ad {
+	const struct bt_data *data;
+	size_t len;
+};
+
+struct ad_stream {
+	/* ad is a two dimensional array of struct bt_data elements. */
+	const struct bt_ad *ad;
+	size_t ad_len;
+
+	/* Indices into the two dimensional array of elements. */
+	size_t ad_index;
+	size_t data_index;
+
+	uint8_t current_data_offset;
+};
+
+void ad_stream_new(struct ad_stream *stream, const struct bt_ad *ad, size_t ad_len)
+{
+	(void)memset(stream, 0, sizeof(*stream));
+	stream->ad = ad;
+	stream->ad_len = ad_len;
+}
+
+bool ad_stream_is_empty(const struct ad_stream *stream)
+{
+	return stream->ad == NULL;
+}
+
+/* Read data from the flattened AD stream. */
+uint8_t ad_stream_read(struct ad_stream *stream, uint8_t *buf, uint8_t *len)
+{
+	uint8_t operation;
+	const size_t ad_index = stream->ad_index;
+	const size_t data_index = stream->data_index;
+	const struct bt_data *current_data = &stream->ad[ad_index].data[data_index];
+	const bool is_first_ltv = (ad_index == 0) && (data_index == 0);
+	const bool is_last_ltv = (ad_index == stream->ad_len - 1) &&
+				  (data_index == stream->ad[ad_index].len - 1);
+
+	if (stream->current_data_offset == 0) {
+		/* Start reading out a new bt_data structure */
+
+		const size_t max_data_to_be_read = current_data->data_len + 2;
+
+		buf[0] = current_data->data_len + 1;
+		buf[1] = current_data->type;
+
+		if (max_data_to_be_read <= *len) {
+			/* The entire bt_data structure fits within the buffer */
+			*len = max_data_to_be_read;
+		} else {
+			/* We can only fit parts of the data structure, the length is not updated */
+		}
+
+		(void)memcpy(&buf[2], current_data->data, max_data_to_be_read - 2);
+		stream->current_data_offset = max_data_to_be_read - 2;
+
+	} else {
+
+		/* Read the remaining data */
+		const ssize_t max_data_to_be_read =
+				current_data->data_len - stream->current_data_offset;
+
+		if (max_data_to_be_read <= *len) {
+			/* The entire bt_data structure fits within the buffer */
+			*len = max_data_to_be_read;
+		} else {
+			/* We can only fit parts of the data structure, the length is not updated */
+		}
+
+		(void)memcpy(&buf[0], current_data->data, max_data_to_be_read);
+		stream->current_data_offset += max_data_to_be_read;
+	}
+
+	if (is_first_ltv && is_last_ltv &&
+	    stream->current_data_offset == current_data->data_len) {
```

Comment:
The comparison`stream->current_data_offset == current_data->data_len` is done multiple times. 

It is still a bit unclear to me what this exactly means, so perhaps storing the result of this in a variable could be beneficial.

Reply:
Renamed `current_data` -> `current_ltv` to make it more clear